### PR TITLE
Decouple the write handler from direct interactionmodelengine usage

### PR DESCRIPTION
**Pull request by andy31415**
IMEngine grew too much and creates very strong coupling between IMEngine and WriteHandler.

This changes the include from this:

image

To this:

image

### Changes

- started to pull out some of the stand-alone functions from `InteractionModelEngine.h` (which seems to have become a catch-all of everything) into a separate `InteractionModelFunctions.h`. It is not great as those functions are very loose as is, however it makes things not include InteractionModelEngine large header anymore
- Added a `Delegate` member to WriteHandler. This may increase RAM usage a bit, however we gain decoupling and testability.

### CODE REVIEW

File: src/app/InteractionModelFunctions.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace chip {
+namespace app {
+
+class WriteHandler;
+
+/**
+ * Returns the metadata of the attribute for the given path.
+ *
+ * @retval The metadata of the attribute, will return null if the given attribute does not exists.
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelFunctions.h:34]:
```suggestion
 * @retval The metadata of the attribute, will return null if the given attribute does not exist.
```

File: src/app/WriteHandler.h [hunk 2]
```diff
@@ -35,6 +35,21 @@
...
+
+class WriteHandler;
+
+class WriteHandlerDelegate
+{
+public:
+    virtual ~WriteHandlerDelegate() = default;
+
+    /**
+     * Returns whether the write operation to the given path is conflict with another write operations. (i.e. another write
```
Comment (bzbarsky-apple) [inline src/app/WriteHandler.h:47]:
```suggestion
     * Returns whether the write operation to the given path is conflicting with another write operation (i.e. another write
```

File: src/app/WriteHandler.h [hunk 3]
```diff
@@ -35,6 +35,21 @@
...
+     * transaction is in the middle of processing the chunked value of the given path.)
```
Comment (bzbarsky-apple) [inline src/app/WriteHandler.h:48]:
```suggestion
     * transaction is in the middle of processing the chunked value of the given path).
```

**Comment by andy31415** [comment]
To many updates and time passed. Will have to re-work this one if we get to more decoupling work.

**Pull request closed without merging**